### ISSUE TITLE
Amazon Linux NameError: global name '__salt__' is not defined

### ISSUE THREAD

**Issue by mendeda0617** [saltstack/salt#27169]
I am seeing the same behavior below with the Salt Minion installed through Amazon Linux as recently as today.   As with the original post of this issue, I receive the same errors for any calls using the following states with a minion at v2015.5.5 :

     pkg.installed
     user.present

Errors reported for each one are below:
```
Function: user.present
Result: False
Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.6/site-packages/salt/state.py", line 1561, in call
                  **cdata['kwargs'])
                File "/usr/lib64/python2.6/contextlib.py", line 34, in __exit__
                  self.gen.throw(type, value, traceback)
                File "/usr/lib/python2.6/site-packages/salt/utils/context.py", line 42, in func_globals_inject
                  yield
                File "/usr/lib/python2.6/site-packages/salt/state.py", line 1561, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.6/site-packages/salt/states/user.py", line 504, in present
                  createhome=createhome):
                File "/usr/lib/python2.6/site-packages/salt/modules/useradd.py", line 161, in add
                  ret = __salt__['cmd.run_all'](cmd, python_shell=False)
              NameError: global name '__salt__' is not defined


Function: pkg.installed
Result: False
Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.6/site-packages/salt/state.py", line 1561, in call
                  **cdata['kwargs'])
                File "/usr/lib64/python2.6/contextlib.py", line 34, in __exit__
                  self.gen.throw(type, value, traceback)
                File "/usr/lib/python2.6/site-packages/salt/utils/context.py", line 42, in func_globals_inject
                  yield
                File "/usr/lib/python2.6/site-packages/salt/state.py", line 1561, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.6/site-packages/salt/states/pkg.py", line 871, in installed
                  **kwargs)
                File "/usr/lib/python2.6/site-packages/salt/states/pkg.py", line 243, in _find_install_targets
                  cur_pkgs = __salt__['pkg.list_pkgs'](versions_as_list=True, **kwargs)
                File "/usr/lib/python2.6/site-packages/salt/modules/yumpkg.py", line 572, in list_pkgs
                  for pkginfo in _repoquery_pkginfo('--all --pkgnarrow=installed'):
                File "/usr/lib/python2.6/site-packages/salt/modules/yumpkg.py", line 129, in _repoquery_pkginfo
                  for line in _repoquery(repoquery_args):
                File "/usr/lib/python2.6/site-packages/salt/modules/yumpkg.py", line 160, in _repoquery
                  call = __salt__['cmd.run_all'](cmd, output_loglevel='trace')
              NameError: global name '__salt__' is not defined

```
I can repeat this behavior with any Salt Minion installed on Amazon Linux that is running v 2015.5.5.  However, my Salt Minions that are using version 2014.7.5 work without issues using the same Salt Master and SLS files.  My Salt Master is version 2015.5.5.  

Thanks,
Dave

**Comment by jfindlay** [comment]
@mendeda0617, thanks for the report.  Are these new minions or upgraded minions?  Is it possible to test with a new minion if not?  Does this failure happen only for those two state functions?

**Comment by mendeda0617** [comment]
It occurs on newly installed minions.  At this time, it is only those two
functions but I am going through more of my SLS files today to see if any
others produce the same error message.

**Comment by jfindlay** [comment]
Thanks for the extra info, @mendeda0617.  It seems likely that this is an error with the loader, since the `__salt__` dictionary is generated by the loader.  Minion debug logs would likely be useful in tracing what that error is.  This dictionary is also used in execution modules, so the problem should also be triggered by something like `salt minion pkg.install htop`.

**Comment by cachedout** [comment]
Are you using any customized modules or grains?

**Comment by mendeda0617** [comment]
No, I am not using any customized modules or customized grains in the
environment.

**Comment by derek0377** [comment]
I meet that problem before, In my scene when minions take the module_refresh,code like this:
             self.functions = salt.loader.minion_mods(self.opts)
            self.returners = salt.loader.returners(self.opts, self.functions)

https://github.com/saltstack/salt/blob/develop/salt/minion.py#L1303

It will refresh  modules's __salt__ variable, 
and at same time, execute modules will meet the __salt__ variable can not be found

**Comment by thehunmonkgroup** [comment]
Also experiencing this on Debian Jessie, installing both v2015.5.5 and v2015.8.0 using git option with the bootstrap script. Downgrading to v2014.7.6 the problem does not occur.

**Comment by cachedout** [comment]
@thehunmonkgroup Wow, interesting. Sounds like this is affecting multiple platforms then. I'm going to see if I can build a Debian 8 box that has this problem. I assume this was a fresh install and not an upgrade? Also, can you post a `salt --versions-report` so I can see what libs are installed? Thanks.

**Comment by thehunmonkgroup** [comment]
@cachedout fresh install on debian 8.2 bento base box with v2015.8.1, and this problem no longer occurs for me. didn't try latest v2015.5.x

**Comment by cachedout** [comment]
I suspect this may be fixed by #29065. Could anybody whom has experienced this problem please verify?

**Comment by cachedout** [comment]
If nobody can verify this, we'll need to close this soon. If there is anybody who has seen this issue who can verify a fix, please leave a comment. Thanks.

**Comment by mendeda0617** [comment]
Hi Mike,

I just installed a fresh copy of v2015.5.5 that is the latest version
available for Amazon Linux using the standard bootstrap install process
(wget -O - http://bootstrap.saltstack.org | sudo sh).  I am still seeing
the same error with that version and its the latest one available for that
process.

Do you need me to install the latest v2015.8.1 from git and test as well?

Thanks,
Dave

**Comment by rallytime** [comment]
@mendeda0617 Apologies for the slow response. This particular fix should be available in the latest release on the 2015.8 branch, 2015.8.3. You can install on Amazon Linux using the v2015.8.3 tag, as you suggested above.

**Comment by mendeda0617** [comment]
Hi Nicole,

I can verify that the fix is working properly in v2015.8.3 on Amazon
Linux.   I am able to now execute the same SLS files on the minions and do
not experience the error in this ticket any longer.   Thanks to the team
for fixing this!
​
Regards,
Dave

**Comment by rallytime** [comment]
@mendeda0617 Great news! Thank you for testing and confirming. I'll go ahead and close this.
